File: testkit/src/main/resources/scenarios/java/ProductId.java
package scanarios.java;

import com.qualityminds.lazyval.LazyValue;
import util.IdGenerator;

import java.util.UUID;

@LazyValue
public record ProductId(String value) {

    public static ProductId of(String value){
        if(value == null){
            return null;
        }
        return new ProductId(value);
    }

    public static ProductId createNew(IdGenerator generator){
        return new ProductId(generator.generateId());
    }

    public static ProductId createNew(){
        return new ProductId(UUID.randomUUID().toString());
    }
}

File: testkit/src/main/resources/scenarios/kotlin/ProductId.kt
package scenarios.kotlin

import com.qualityminds.lazyval.LazyValue
import java.util.UUID
import util.IdGenerator

@LazyValue
class ProductId private constructor(val value: String) {
    companion object {
        @JvmStatic
        fun of(value: String): ProductId {
            require(value.isNotBlank()) { "ProductId cannot be blank" }
            return ProductId(value)
        }

        @JvmStatic
        fun createNew(): ProductId {
            return ProductId(UUID.randomUUID().toString())
        }

        @JvmStatic
        fun createNew(generator: IdGenerator): ProductId {
            return ProductId(generator.generateId())
        }
    }
}